### PR TITLE
(WIP) Bitbucket Token Refresh

### PR DESCRIPTION
**Pull request by zanedev**
**- Summary**
Added support for automatic refreshing the bitbucket oauth token using the api provided refresh token and expiry date.

NOTE: There are a few areas I wasn't sure about the code organization.  I put //todos everywhere I had questions/concerns.  Please let me know if anybody has any feedback.

Outstanding Issues:
- The auth base_url is hard coded currently to an instance of the custom oauth server at zeit because I am not sure of the best way to pass it down from the App props used in the initial auth code.
- The refresh token is stored client side which is a potential security issue.  (dev meeting 8/30: this is a low risk issue so we will go ahead for now and replace it later when we have a server side refresh token implementation)
- The auth refresh token code requires CORS enabled on the auth server because it requests it directly from the browser fetch rather than going through the browser window callback pattern used with direct authentication. (dev meeting 8/30: this is okay and supported on netlify side)
- In it's current form the refresh token will not work with netlify auth server because it doesn't have the token exchange logic.  (dev meeting 8/30: @biilmann will look into whether it's quick to add a simple token exchange endpoint for netlify auth server, if not we will launch with only custom oauth server support meaning the user would be logged out automatically every hour and some code will need to be added to handle this case of not requesting a new token if using netlify auth) @biilmann for reference [here is the basic token exchange logic](https://github.com/zanedev/netlify-cms-bitbucket-oauth-provider/blob/88c9c7c78ea0aea006470e829c6c3901fdf96b5d/index.js#L40) needed to keep it working with netlify auth server: 
- Will add more testing before we merge.

**- Test plan**
To test use the [custom bitbucket oauth server](https://github.com/zanedev/netlify-cms-bitbucket-oauth-provider) because netlify auth doesn't support the refresh token yet.  Change the bitbucket/api.isTokenExpired function to return true to always request the new token and observe it does indeed get a new token with each request and assign it to the user.  Alternatively, do not change the function and wait for the timeout (1hr) (or lower the timeout) and see that the app continues to work after the timeout and automatically refreshes the token rather than getting 401 responses and freezing up like it used to do.

**- Description for the changelog**
Added automatic bitbucket api auth refresh token logic

**- A picture of a cute animal (not mandatory but encouraged)**
ricky-kharawala-10194
Photo by Ricky Kharawala on Unsplash

### CODE REVIEW

File: src/lib/netlify-auth.js [hunk 1]
```diff
@@ -74,6 +74,25 @@ class Authenticator {
     return host === 'localhost' ? null : host;
   }
 
+  //NOTE the auth server needs to have CORS enabled for this direct request to work
+  //todo: is this the best place for this refresh auth token code to live?
+  //todo: will netlify auth server use cors and direct request?
+  refreshAuthToken(options) {
+    // const base_url = this.base_url; //todo: how to get proper config base_url from app/props config.base_url
+    const base_url = "https://netlify-bitbucket-oauth.now.sh";
```
Comment (tech4him1) [inline src/lib/netlify-auth.js]: Can you just use `this.base_url` here like the other functions do (created at line 34)?

**Comment by zanedev** [comment]
You'd think so but it's passed down through app prop types in the functions
above so I'm not sure how to access the original prop value without saving
it globally or something...

-- Zane

**Comment by tech4him1** [comment]
Sorry, I noticed that after I posted. It would probably have to be sent down through `actions/auth.js` or something with the initial config.

**Comment by tech4him1** [comment]
Quoting @biilmann in Gitter ([January 9, 2018 12:58 PM](https://gitter.im/netlify/NetlifyCMS?at=5a551ee383152df26d4f39fa)):
> I don’t think we’ll be adding more endpoints to auth.netlify.com - also since I’m not toally sure it rhymes with GitLabs security model to distribute the refresh token and give an endpoint where you can use it without having the client secret

> So I think the current logic of manually loging in again is better if using the auth.netlify.com setup - and then the Git Gateway can be smarter about it and do automatic refresh token handling

> since there we wont share the git tokens with the client at all

**Comment by erquhart** [comment]
Closing in favor of #1504.

**Pull request closed without merging**